File: java/src/main/java/com/github/algo/backtracking/WordSearch.java
package com.github.algo.backtracking;

import org.javatuples.Pair;

import java.util.HashSet;
import java.util.Set;

public class WordSearch {

    int[][] movements = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};

    private boolean dfs(int i, int j, int c, char[][] board, Set<Pair<Integer, Integer>> visited, String word) {
        if (c >= word.length()) {
            return true;
        }
        visited.add(Pair.with(i, j));
        for (int[] move : movements) {
            int di = i + move[0], dj = j + move[1];
            if (di >= 0 && di < board.length && dj >= 0 && dj < board[i].length && !visited.contains(Pair.with(di, dj))) {
                if (word.charAt(c) == board[di][dj] && dfs(di, dj, c + 1, board, visited, word)) {
                    return true;
                }
            }
        }
        visited.remove(Pair.with(i, j));
        return false;
    }

    public boolean exist(char[][] board, String word) {
        Set<Pair<Integer, Integer>> visited = new HashSet<>();
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[i].length; j++) {
                if (board[i][j] == word.charAt(0) && dfs(i, j, 1, board, visited, word)) {
                    return true;
                }
            }
        }
        return false;
    }
}

File: python/main/graphs/WordSearch.py
from typing import List

MOVEMENTS = [(1, 0), (0, 1), (-1, 0), (0, -1)]


class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        visited = set()

        def dfs(i, j, wi):
            if wi >= len(word):
                return True
            visited.add((i, j))
            for di, dj in MOVEMENTS:
                ni = i + di
                nj = j + dj
                if 0 <= ni < len(board) and 0 <= nj < len(board[ni]):
                    if (ni, nj) not in visited and word[wi] == board[ni][nj]:
                        if dfs(ni, nj, wi + 1):
                            return True
            visited.remove((i, j))
            return False

        for i in range(len(board)):
            for j in range(len(board[i])):
                if board[i][j] == word[0]:
                    if dfs(i, j, 1):
                        return True
        return False

        pass


if __name__ == "__main__":
    solution = Solution()
    output = solution.exist([["C", "A", "A"], ["A", "A", "A"], ["B", "C", "D"]], "AAB")
    print(output)
